plugins/orchestrator/orchestrator.go

```diff
@@ -0,0 +1,187 @@
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at:
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package orchestrator
+
+import (
+	"github.com/ligato/cn-infra/datasync"
+	"github.com/ligato/cn-infra/infra"
+	"github.com/ligato/cn-infra/rpc/grpc"
+	api "github.com/ligato/vpp-agent/api/genericmanager"
+	"github.com/ligato/vpp-agent/pkg/models"
+	kvs "github.com/ligato/vpp-agent/plugins/kvscheduler/api"
+	"golang.org/x/net/context"
+)
+
+// Plugin implements sync service for GRPC.
+type Plugin struct {
+	Deps
+
+	manager *genericManagerSvc
+
+	// datasync channels
+	changeChan   chan datasync.ChangeEvent
+	resyncChan   chan datasync.ResyncEvent
+	watchDataReg datasync.WatchRegistration
+
+	*dispatcher
+}
+
+// Deps represents dependencies for the plugin.
+type Deps struct {
+	infra.PluginDeps
+
+	GRPC        grpc.Server
+	KVScheduler kvs.KVScheduler
+	Watcher     datasync.KeyValProtoWatcher
+}
+
+// Init registers the service to GRPC server.
+func (p *Plugin) Init() error {
+	p.dispatcher = &dispatcher{
+		log:   p.Log.NewLogger("dispatcher"),
+		store: newMemStore(),
+		kvs:   p.KVScheduler,
+	}
+
+	// initialize datasync channels
+	p.resyncChan = make(chan datasync.ResyncEvent)
+	p.changeChan = make(chan datasync.ChangeEvent)
+
+	// register grpc service
+	p.manager = &genericManagerSvc{
+		log:      p.log,
+		dispatch: p.dispatcher,
+	}
+
+	if grpcServer := p.GRPC.GetServer(); grpcServer != nil {
+		api.RegisterGenericManagerServer(grpcServer, p.manager)
+	} else {
+		p.log.Infof("grpc server not available")
+	}
+
+	return nil
+}
+
+// AfterInit subscribes to known NB prefixes.
+func (p *Plugin) AfterInit() (err error) {
+	go p.watchEvents()
+
+	nbPrefixes := p.kvs.GetRegisteredNBKeyPrefixes()
+	if len(nbPrefixes) > 0 {
+		p.log.Infof("watch starting for %d registered NB prefixes", len(nbPrefixes))
+	} else {
+		p.log.Warnf("no registered NB prefixes found in KVScheduler")
+	}
+
+	var prefixes []string
+	for _, prefix := range nbPrefixes {
+		//prefix = path.Join("config", prefix)
+		p.log.Debugf("- watching NB prefix: %s", prefix)
+		prefixes = append(prefixes, prefix)
+	}
+
+	p.watchDataReg, err = p.Watcher.Watch(p.PluginName.String(),
+		p.changeChan, p.resyncChan, prefixes...)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// InitialSync will start initial synchronization with downstream.
+func (p *Plugin) InitialSync() {
+	p.Log.Debugf("starting initial sync")
+	txn := p.KVScheduler.StartNBTransaction()
+	ctx := kvs.WithResync(context.Background(), kvs.DownstreamResync, true)
+	if _, err := txn.Commit(ctx); err != nil {
```

Comment:
Will this not remove everything configured after agent restart?

Reply:
Is there a way to sync with SB without removing everything configured?? I was under impression that:
- FullResync: syncs NB-KVS-SB (comparing NB-SB and fixing differences)
- DownstreamResync: syns KVS-SB (not touching anything in SB, just dumping to update)
- UpstreamResync: syncs NB-KVS? (not sure here)